### DISCUSSION TITLE
Nested taxonomy links?

### DISCUSSION THREAD

**Discussion by stephenmeehanuk** [statamic/cms#5790] [General]
Hi, 

I've got a collection called `products`, I've attached four `taxonomies` to the collection.

I'd like to create a sidebar menu, listing terms from each taxonomy, but I can't figure out how to display the taxonomies so they are nested?

This outputs a long list alphabetically, of all the terms from each of the four taxonomies

```
<ul >
    {{ taxonomy from="group_1|group_2|group_3|group_4" }}
        <li><a href="{{ url }}">{{ title }}</a></li>
    {{ /taxonomy }}
</ul>
```

This gives me a flat list, ordered alphabetically...

```
- Group 1 Term
- Group 2 Term
- Group 3 Term 
- Group 1 Term 
- Group 2 Term 
- Group 2 Term 
- Group 1 Term 
- Group 1 Term 
```

What I want to do is nest them, to create a taxonomy menu like this

```
- Group 1 Term 
	- Group 2 Term 
		- Group 3 Term 
- Group 1 Term 
	- Group 2 Term 
	- Group 2 Term 
- Group 1 Term 
- Group 1 Term 	
```

I think I understand why the `taxonomy` works like this. It’s simply spitting out all the terms from each taxonomy - it has no way to know if they are related...

Perhaps the `collection` tag is a better route? I’m using the four taxonomies on a collection called products. Here each entry is associated with more than one taxonomy. Could this be used to generate a nested menu?

```
{{ collection:products }}
    ...
{{ /collection:products }}

```

I'm thinking I need to loop through `collection:products`, instead of displaying entries - is there a way to list nested taxonomies based on their relationship with an entry?

I'm considering building a navigation, but that'll mean manually adding the name and url of each term, not ideal.

Is there a way to achieve this using antlers?

```
Statamic 3.3.5 Pro
Laravel 9.5.1
PHP 8.0.8
Parser: Regex (default)
```

**Comment by jasonvarga** [comment]
Unless I'm misunderstanding, it sounds like you could use nested taxonomy tag loops.

```mustache
<ul>
{{ taxonomy from="one" }}
  <li>
    {{ title }}
    <ul>
    {{ taxonomy from="two" }}
        <li>
            {{ title }}
            <ul>
            {{ taxonomy from="three" }}
                <li>{{ title }}</li>
            {{ /taxonomy }}
            </ul>
        </li>
    {{ /taxonomy }}
    </ul>
  </li>
{{ /taxonomy }}
</ul>
```

**Reply by stephenmeehanuk** [reply]
Thanks, but nesting like that gives me this error?
```
preg_replace(): Compilation failed: regular expression is too large at offset 42201
```
When I reduce the nesting to two levels, it takes about 30 seconds to load and doesn't nest as expected. There's a lot of repetition, like it's looping over itself over and over.

The screenshot below shows the `products` collection. I'm trying to build a sidebar menu that uses `group 1` as the top level menu,  the following groups, 2, 3 and 4 need to be nested based on the relationship with each other - as shown below. Is this possible?

![Screenshot 2022-04-08 at 08 58 25](https://user-images.githubusercontent.com/72254741/162391848-028871c6-1e9e-4dfd-abd3-636a06b0030c.png)

I've created a test navigation in Statamic, which shows what I'm trying to make.

![Screenshot 2022-04-08 at 09 12 11](https://user-images.githubusercontent.com/72254741/162394191-8b1ef993-eaaf-4e1e-9703-ef8bbee329ed.png)

If navigation could support adding taxonomies in the same way as entires, this would be trivial to create. However, using navigation for this isn't ideal as the names and urls are hardcoded in the navigation.

Is it possible to create a nested taxonomy based navigation in antlers like the example above? 

Or better yet, could a (near) future Statamic update support taxonomies in navigation?

**Reply by jasonvarga** [reply]
See https://github.com/statamic/ideas/issues/631

As for the error, make sure you're using the new parser.

**Reply by stephenmeehanuk** [reply]
> See https://github.com/statamic/ideas/issues/631
I forgot I ran into this same issue last summer 😃 . It'd be great if navigation could support taxonomies. Any plans to do so?

**Comment by afonic** [comment]
This looks like a bit too complex data structure design. Your "Products" sound like product categories. Maybe you're better off creating a collection for each?

I'm trying to understand why you have 4 taxonomy fields (Groups 1 to 4) and not one that can select terms for each four. If you explain it better maybe we can help you create a custom tag that does what you're trying to do.

**Reply by stephenmeehanuk** [reply]
Thanks for the reply

Products are just that, `collection:products` of 105 entries. I did have a taxonomy called `solutions`, this contained all the terms but I couldn't find a way to nest the terms.  I'm trying to build a sidebar navigation, that lists all the terms and sub terms. But it's proving quite tricky.

For example, I have a product called `Containment Walls` It's a `Precast Structure` product, but it also needs to be tagged as an `Underground Structure` product.

I've used multiple taxonomies because we can't (yet) nest terms. 

I think I might have a work around. I can relate the other taxonomies by adding an `entires` field to each taxonomy blueprint.

Screenshot below shows this
![Screenshot 2022-04-08 at 11 45 12](https://user-images.githubusercontent.com/72254741/162420812-cffef2ef-0989-41e7-a4d6-5307c470699f.png)

Using this I can display nested taxonomies, I've removed the HTML from this example.
```
{{ taxonomy slug:is="{slug}" from="group_1" collection="{segment_1}" }}    

    {{ group_2_taxonomy_terms_field }}
        {{ title }}
            {{ entries}}  
                {{ title }}
            {{ /entries }} 
                
                {{ group_3_taxonomy_terms_field  }}
                    {{ title }}
                        {{ entries}}  
                            {{ title }}
                        {{ /entries }} 
                {{ /group_3_taxonomy_terms_field }}

    {{ /group_2_taxonomy_terms_field }}
        
{{ /taxonomy }}
```
However, because each entry has more than one taxonomy term there's unwanted duplication. The `Containment Walls` entry has two taxonmies, `Precast Structures` and `Underground Structures`

![Screenshot 2022-04-08 at 12 28 04](https://user-images.githubusercontent.com/72254741/162426960-211c2b9d-bd42-4e5e-a568-aa5652324000.png)


For this particular menu, I'd like to exclude any entires from `group_1` that are also in `group_2`, `group_3` or (gulp) `group_4`

**Comment by stephenmeehanuk** [comment]
Managed to get this working... 

```
{{ taxonomy slug:is="{slug}" from="group_1" collection="{segment_1}" }}    

    {{ group_2_taxonomy_terms_field }}
        {{ title }}
            {{ entries}}  
            {{ if group_2 & !group_3 }} // Only show entires from group_2
                {{ title }}
              {{ /if }}  
            {{ /entries }} 
                
                {{ group_3_taxonomy_terms_field  }}
                    {{ title }}
                        {{ entries}}  
                     {{ if group_3 & !group_4 }}  // Only show entires from group_3
                            {{ title }}
                              {{ /if }}  
                        {{ /entries }} 
                {{ /group_3_taxonomy_terms_field }}

    {{ /group_2_taxonomy_terms_field }}
        
{{ /taxonomy }}
```

![Screenshot 2022-04-08 at 14 22 11](https://user-images.githubusercontent.com/72254741/162444426-1f63a2aa-eb5d-43f9-9c32-b3c0fd687177.png)

Writing the steps out here for future me.

1. Assign terms to each `products` entry - this needs to be done regardless
2. Extra step: Within each taxonomy, add related taxonomies
